### ISSUE TITLE
BUG: Array (and scalar) `dealloc` slow in 1.17 (previously only if memoryviews of arrays were made)

### ISSUE THREAD

**Issue by ttrd** [numpy/numpy#14701]
I noticed that slicing arrays in 1.17.2 takes 20-30% more time than it takes in 1.16.4. Is that already known resp. accepted?

### Reproducing code example:

```python
from timeit import timeit
 
setup = '''
import numpy as np
arraySize = 10**6
values = np.random.randint(0, 255, (arraySize,))
'''

stmt = '''
for i in range(arraySize):
    x = values[i:]
'''
 
number = 100
 
total = timeit(stmt=stmt, setup=setup, number=number)
print(f'{total/number:.6e}')
```

Example output using numpy 1.16.4: 1.387783e-01
Example output using numpy 1.17.2: 1.785848e-01

### Numpy/Python version information:

1.16.4 3.7.3 (v3.7.3:ef4ec6ed12, Mar 25 2019, 22:22:05) [MSC v.1916 64 bit (AMD64)]
1.17.2 3.7.3 (v3.7.3:ef4ec6ed12, Mar 25 2019, 22:22:05) [MSC v.1916 64 bit (AMD64)]

**Comment by eric-wieser** [comment]
Any chance you could use the command line form of timeit which includes the variance?

**Comment by ttrd** [comment]
Like this?

```
python -m timeit -n 100 -s "import numpy as np; arraySize=10**6; values=np.random.randint(0, 255, (arraySize,))" "for i in range(arraySize): x=values[i:]"
```

numpy 1.16.4: 100 loops, best of 5: 127 msec per loop
numpy 1.17.2: 100 loops, best of 5: 158 msec per loop

I could not find an option to include the variance in the output. However, in the documentation of timeit, it is stated that it is tempting to calculate mean and standard deviation, but the minimum is more useful.

**Comment by eric-wieser** [comment]
Hmm, I guess I'm thinking of the ipython `%timeit`. Maybe you're right though, and the variance isn't interesting.

**Comment by seberg** [comment]
I bisected it down to this commit: fa8af41c9d375072b2c7af66e5f7f01df4754841; After which, confused, I commented out the `import random` in `__init__.py`. And the final conclusion is: Removing the random import changes the speed of this example (just use `ones` instead of `random`). Adding `gc.disable()` to the setup does not seem to change it. I have no idea how random could affect this.

**Comment by seberg** [comment]
OK, I tracked this down. And maybe it actually would make sense to fix it. There is no real performance regression here, in the sense that it was always slow if, and only if, the buffer interface was at some point used with an array that is still alive. `random` seems to cause this to happen:

```
In [1]: arr = np.empty(3)                                                                                                     
In [2]: %timeit arr[...]                                                                                                      
160 ns ± 0.0494 ns per loop (mean ± std. dev. of 7 runs, 10000000 loops each)

In [3]: memoryview(arr)
# the above causes a global dictionary to be filled, which requires lookup on every array deallocation

[4]: %timeit arr[...]                                                                                                      
190 ns ± 1.27 ns per loop (mean ± std. dev. of 7 runs, 10000000 loops each)
```

So the difference is between whether or not the global buffer cache is empty or not. If we are good with with limiting the size of the cache drastically, we could use a much faster LRU cache, it might even be better to limit it?

**Comment by seberg** [comment]
I just remembered this, may add it to triage. It is one of those small, but a bit annoying things. However, a real fix is non-trivial probably.

So that my thoughts and starts on it do not get lost some notes here:

1. This also affects scalars, so that the performance of `one = np.float64(1)` `%timeit one + one` decreases by about 30%
   * The scalar regression is also a bit newer, because previously there was a memory leak here.
2. This affects any array destruction, and even short-lived views. For people handling small arrays, this is a slight penalty, although unlike scalars, there are other overheads that probably hide it more.

Now on actually fixing it: This is tricky due to the buffer PEP being a bit unclear on who owns the memory. If the buffer PEP would be modified to say that the data pointed to must be held by the original object *while the buffer information may already be cleared up*, the current issue of `"s"` parsing in `PyArg_Parse...` that prompts the need for the workaround would go away I believe.

Assuming that fixing up-stream is difficult this **can** be fixed for scalars (except void):
  * Scalars always have the exact same buffer information, this means the buffer information could either point to static and constant data, or the cache set up to always find the same information for all basic scalars.
  * Void scalars have more complex buffer information, and must be treated like arrays.

A full fix for arrays is complicated, this can basically only be done by using something like weak references. One could:
  1. Use weak-references as is, the issue is that their C-API is a bit limited, which means python calls will occur when an array is deallocated (and a memoryview of it created at some point), this may be OK.
  2. Use weak-references but use Python C-API that is considered private.
  3. Create a new field in all array objects which holds our own tiny implementation of weak references.

No matter which way, void scalars would have to get a new field, but they already have a few fields so that seems fine. In the path of using weak references, VOID scalars would be the only ones supporting them (but that seems fine).

There may be other approaches, but I am not sure about it. It could be a project for someone willing to fight with python a bit. **It probably makes sense to separate out a fix for normal scalars, since they seem to have a noticable speed impact due to this.** Fixing only the scalars should only be moderately difficult.

**Comment by seberg** [comment]
This has been addressed in NumPy 1.20 with the PRs gh-16938 and gh-16389.
